src/catalua.h

```diff
@@ -35,6 +35,21 @@ int lua_mapgen( map *m, const oter_id &terrain_type, const mapgendata &md, int t
 void lua_callback( const char *callback_name );
 
 /**
+ * Execute a callback that can be overriden by all mods with 1 argument(s) put to global table (_G).
+ */
+void lua_callback( const char *callback_name, const char *callback_arg1 );
+
+/**
+ * Execute a callback that can be overriden by all mods with 2 argument(s) put to global table (_G).
+ */
+void lua_callback( const char *callback_name, const char *callback_arg1, const char *callback_arg2 );
```

Comment:
Variables / parameters with numbers in them are a code smell - use a vector or similar instead. You don't need n callback functions, only one generic.

Even better would be a function template, so the caller doesn't need to convert everything into a string, which the Lua code has to convert back into whatever it actually is.

Reply:
> Variables / parameters with numbers in them are a code smell - use a vector or similar instead. You don't need n callback functions, only one generic.

> Even better would be a function template, so the caller doesn't need to convert everything into a string, which the Lua code has to convert back into whatever it actually is.


I've tried using `cstdarg` and its `...`, but didn't achieve good results - if I understood it right, all of the parameters should be of the same type, but it will be required to have multiple types in a single callback.

I've decided to use string for the time being as it nicely fits for various numbers and string identifiers, but I will definitely need to rework this as I will need references to certain objects.

I didn't work with function templates previously, but I beleive it is something about preprocessor magic, right?